### PR TITLE
Updated the CLI to accept parameters for the nnUNet models. Additional stuff for packaging etc.

### PR DESCRIPTION
**Pull request by georgeaidinis**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -27,13 +38,13 @@
...
 		'pymongo==3.12.0',
-		'scikit-learn==0.24.2',
+		'scikit-learn',
 		'nibabel==3.2.1',
 		'resource==0.2.1',
 		'networkx==2.5.1',
 		'pandas==1.2.5',
 		'pathlib',
-                'DeepMRSeg @ git+https://github.com/CBICA/DeepMRSeg.git@main'
+        'DeepMRSeg @ git+https://github.com/CBICA/DeepMRSeg.git@main'
```
Comment (ashishsingh18) [inline setup.py:47]: This should be removed since we are going to nnunet version.

File: requirements.txt [hunk 2]
```diff
@@ -1,18 +1,106 @@
...
+batchgenerators==0.25
+cachetools==5.3.1
+certifi==2023.5.7
+charset-normalizer==3.1.0
+ci-info==0.3.0
+click==8.1.3
+cloudpickle==2.2.1
+cycler==0.11.0
+decorator==4.4.2
+DeepMRSeg @ git+https://github.com/CBICA/DeepMRSeg.git@f351affff0d79a134fbc4073582f798efc81e409
```
Comment (ashishsingh18) [inline requirements.txt:13]: This must be removed.

File: requirements.txt [hunk 3]
```diff
@@ -1,18 +1,106 @@
...
+dicom2nifti==2.4.8
```
Comment (ashishsingh18) [inline requirements.txt:14]: Why is this needed? We are not handling any dicom data. Please remove if not needed.

File: niCHARTPipelines/__main__.py [hunk 4]
```diff
@@ -6,37 +6,281 @@
...
+    optional arguments: 
+        [DLICVMDL]      The filepath of the DLICV model will be. In case the
+        [--DLICVmdl]    model to be used is an nnUNet model, the filepath of
+                        the model's parent directory should be given. Example: 
+                        /path/to/nnUNetTrainedModels/nnUNet/
+        
+        [DLMUSEMDL]     The filepath of the DLMUSE model will be. In case the
+        [--DLMUSEmdl]   model to be used is an nnUNet model, the filepath of
+                        the model's parent directory should be given. Example:
+                        /path/to/nnUNetTrainedModels/nnUNet/
```
Comment (ashishsingh18) [inline niCHARTPipelines/__main__.py:56]: Shouldn't these be removed since both models are now passed in via another param.

File: niCHARTPipelines/Structural.py [hunk 5]
```diff
@@ -3,24 +3,59 @@
...
-import CalculateROIVolumeInterface
+from niCHARTPipelines import nnUNetInterface
+from niCHARTPipelines import MaskImageInterface
+from niCHARTPipelines import ROIRelabelInterface
+from niCHARTPipelines import CalculateROIVolumeInterface
 
-def run_structural_pipeline(inDir,DLICVmdl,DLMUSEmdl,outDir, MuseMappingFile,scanID,roiMappingsFile):
+def run_structural_pipeline(inDir,
+                            DLICVmdl_path,
+                            DLMUSEmdl_path,
```
Comment (ashishsingh18) [inline niCHARTPipelines/Structural.py:13]: Shouldn't these be removed since both models are now passed via another param?

File: niCHARTPipelines/Structural.py [hunk 6]
```diff
@@ -3,24 +3,59 @@
...
+                            outDir, 
+                            MuseMappingFile,
+                            scanID,
+                            roiMappingsFile,
+                            nnUNet_raw_data_base=None,
+                            nnUNet_preprocessed=None,
+                            results_folder=None,
```
Comment (ashishsingh18) [inline niCHARTPipelines/Structural.py:20]: Suggestion: better to rename this to 'model_folder' since it actually contains model. The Results_Folder name seems quite confusing. Internally it can still set the env variable os.environ['RESULTS_FOLDER']

File: niCHARTPipelines/__main__.py [hunk 7]
```diff
@@ -6,37 +6,281 @@
...
+                                 arguments are corresponding to a  nnUNet model 
+                                 (v1 needs this currently).
+
+        [NNUNET_PREPROCESSED]   The filepath of the directory where the 
+        [--nnUNet_preprocessed] intermediate preprocessed data are saved. This
+                                argument is only required if the DLICVMDL and
+                                DLMUSEMDL arguments are corresponding to a
+                                nnUNet model (v1 needs this currently).
+
+        [RESULTS_FOLDER]        THIS IS ONLY NEEDED IF BOTH DLICV AND DLMUSE
```
Comment (ashishsingh18) [inline niCHARTPipelines/__main__.py:70]: Better to rename this to 'MODEL_FOLDER'. Internally it can be used to set the os.environ['RESULTS_FOLDER'] env variable.

**Review by ashishsingh18 — CHANGES_REQUESTED** [review]
@georgeaidinis Thanks for the PR. 
I see that tensorflow/DeepMRSeg are still there. These should be removed since we are moving to nnunet version. Also please see some more minor comments/suggestions.

File: setup.py (hunk 1 above)
Comment (georgeaidinis) [reply setup.py:47]: Agree. I left it there in case somebody wanted to keep the old DeepMRSeg approach, but I realize that the code is commented out for the nipype interface.

File: niCHARTPipelines/__main__.py (hunk 4 above)
Comment (georgeaidinis) [reply niCHARTPipelines/__main__.py:56]: I left it like that since I wanted to include the possibility of someone being able to swap one method for a different one, i.e. use nnUNet for DLICV, and something else for DLMUSE. My thinking here was to be futureproof, let's say if we train a nnUNetV2 model for one but not the other steps.

File: niCHARTPipelines/Structural.py (hunk 5 above)
Comment (georgeaidinis) [reply niCHARTPipelines/Structural.py:13]: My thinking here was to be futureproof, let's say if we train a nnUNetV2 model for one but not the other steps.

File: niCHARTPipelines/__main__.py (hunk 7 above)
Comment (georgeaidinis) [reply niCHARTPipelines/__main__.py:70]: Understood, agreed.

**Comment by AlexanderGetka-cbica** [comment]
For migration from the tensorflow base image, we can always use a corresponding pytorch base image:
https://hub.docker.com/r/pytorch/pytorch/tags

I would recommend one of the "runtime" tagged versions for a torch base version that corresponds to what works for the pipelines now (1.12.1? 1.10?)

Alternatively, you could start from a relatively bare "python" image, do the CUDA/cuDNN(?) installations yourself, then install nichartpipelines and see if this minimizes the image.

**Comment by ashishsingh18** [comment]
> 



> For migration from the tensorflow base image, we can always use a corresponding pytorch base image: https://hub.docker.com/r/pytorch/pytorch/tags
> 
> I would recommend one of the "runtime" tagged versions for a torch base version that corresponds to what works for the pipelines now (1.12.1? 1.10?)
> 
> Alternatively, you could start from a relatively bare "python" image, do the CUDA/cuDNN(?) installations yourself, then install nichartpipelines and see if this minimizes the image.

I would recommend starting from pytorch image so we don't need to manually install and test the other dependencies.
But these dockerfile changes can go in another PR.

**Review by AlexanderGetka-cbica — APPROVED** [review]
Looks good to me. 

Feel free to submit Dockerfile changes and additional changes to the container discussed in our last meeting as a separate PR.

**Review by ashishsingh18 — APPROVED** [review]
